ssl/ssl_sess.c

```diff
@@ -1081,58 +1102,76 @@ static void SSL_SESSION_list_remove(SSL_CTX *ctx, SSL_SESSION *s)
     s->prev = s->next = NULL;
 }
 
+/* session cache is already locked by caller */
 static void SSL_SESSION_list_add(SSL_CTX *ctx, SSL_SESSION *s)
 {
-    if ((s->next != NULL) && (s->prev != NULL))
-        SSL_SESSION_list_remove(ctx, s);
+    SSL_SESSION_CACHE *session_cache = ctx->session_cache;
 
-    if (ctx->session_cache_head == NULL) {
-        ctx->session_cache_head = s;
-        ctx->session_cache_tail = s;
-        s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
-        s->next = (SSL_SESSION *)&(ctx->session_cache_tail);
+    if (session_cache == NULL)
+        return;
+
+    SSL_SESSION_list_remove(ctx, s);
+
+    if (session_cache->head == NULL) {
+        session_cache->head = s;
+        session_cache->tail = s;
+        s->prev = (SSL_SESSION *)&(session_cache->head);
+        s->next = (SSL_SESSION *)&(session_cache->tail);
     } else {
-        s->next = ctx->session_cache_head;
+        s->next = session_cache->head;
         s->next->prev = s;
-        s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
-        ctx->session_cache_head = s;
+        s->prev = (SSL_SESSION *)&(session_cache->head);
+        session_cache->head = s;
     }
 }
 
 void SSL_CTX_sess_set_new_cb(SSL_CTX *ctx,
                              int (*cb) (struct ssl_st *ssl, SSL_SESSION *sess))
 {
-    ctx->new_session_cb = cb;
+    CRYPTO_THREAD_write_lock(ctx->lock);
+    CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+    ctx->session_cache->new_session_cb = cb;
+    CRYPTO_THREAD_unlock(ctx->session_cache->lock);
+    CRYPTO_THREAD_unlock(ctx->lock);
```

Comment:
Yes. These callback were originally left in the SSL_CTX (in our original implementation), but were moved to the cache structure for some consistency. However; one only of these callbacks is invoked (regardless of the number of SSL_CTX's sharing the cache), depending on which SSL_CTX is "active" when an SSL_SESSION is added, removed, etc; since the SSL_CTX is passed into the callback. If a list of SSL_CTX's were kept in the cache, then a callback could be invoked for each of them, at which point, having SSL_CTX-specifc callbacks makes sense (but so does having a single callback).

So, I'm kinda "meh" on how the callbacks are stored.